driver-core/src/main/com/mongodb/ConnectionString.java

```diff
@@ -706,6 +706,10 @@ private MongoCredential createCredentials(final Map<String, List<String>> option
         }
 
         if (credential != null && authMechanismProperties != null) {
+            if (credential.getAuthenticationMechanism() == AuthenticationMechanism.MONGODB_IAM && (credential.getUserName() == null
```

Comment:
I think this restriction should be enforced at a lower level.  `IamAuthenticator` would be appropriate I think.  Currently, `Authenticator` implementations are the only places that we examine mechanism properties at all.

Reply:
Done. It's checked in the function `getSessionToken` which is always called when creating the client's final message.